.github/workflows/pull_request.yml

```diff
@@ -1,15 +1,87 @@
+on: pull_request
+run-name: "${{github.event.pull_request.title}} (#${{ github.event.number }})"
+concurrency:
+  group: ${{ github.event.number }}
+  cancel-in-progress: true
+
 jobs:
-  validate:
+  ruby-tests:
+    name: Ruby tests
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - uses: ruby/setup-ruby@v1
+        with:
+          bundler-cache: true
+          ruby-version: '3.0'
+        env:
+          BUNDLE_WITH: 'tests'
+      - uses: actions/cache@v4
+        with:
+          path: |
+            /tmp/iso-*.txt
+            /tmp/similarweb
+          key: tmp-files-${{ hashFiles('/tmp/similarweb/*')}}
+      - name: Validate JSON structure
+        run: bundle exec ruby ./tests/validate-json.rb
+      - name: Validate Categories
+        run: bundle exec ruby ./tests/validate-categories.rb
+      - name: Validate Region codes
+        run: bundle exec ruby ./tests/region-codes.rb
+      - name: Validate file extensions and permissions
+        run: tests/validate-fs.sh
+      - name: Validate Images
+        run: bundle exec ruby ./tests/validate-images.rb
+      - name: Validate SVG
+        run: bundle exec ruby ./tests/svg-lint.rb
+      - name: Validate URL/Domain reachability
+        run: bundle exec ruby ./tests/validate-urls.rb
+        continue-on-error: true
+      - name: Quality Checks
+        run: bundle exec ruby ./tests/quality-checks.rb
+      - name: Validate Ruby scripts
+        run: bundle exec rubocop
+
+  node-tests:
+    name: JavaScript tests
     runs-on: ubuntu-latest
 
     steps:
       - name: Checkout code
         uses: actions/checkout@v4
 
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20'
+
+      - name: Install dependencies
+        run: npm install
+
+      - name: Get modified files
+        id: diff
+        run: |
+          echo "::debug:: Fetching files from ${{ github.api_url }}/repos/${{ github.repository }}/pulls/${{ github.event.number }}/files"
+          FILES=$(curl -s "${{ github.api_url }}/repos/${{ github.repository }}/pulls/${{ github.event.number }}/files" | jq -r '.[].filename' | tr '\n' ' ')
+          echo "name=${FILES}" >> $GITHUB_OUTPUT
+          ENTRIES=$(echo "$FILES" | tr ' ' '\n' | grep -E '^entries/.*\.json$' | tr '\n' ' ')
+          if [ -n "$ENTRIES" ]; then
+            echo "entries=${ENTRIES}" >> $GITHUB_OUTPUT
+          fi
+
+      - name: Validate Language codes
+        if: steps.diff.outputs.entries
+        run: node tests/languages.js ${{ steps.diff.outputs.entries }}
+
+  external-tests:
+    name: External Tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
       - name: Call PR Validator
         run: |
           curl -s --fail-with-body "https://pr-validator.2factorauth.workers.dev/${{ github.event.repository.name }}/${{ github.event.number }}/" \
```

Comment:
Should we use the `workers.dev` URL or the subdomain on `2fa.directory`?

Reply:
Whoops! Yeah the `2fa.directory` domain is the preferred one because the WAF can be used then. Fixed in d9ec537